File: sdk/go/http.go
package protocol7h3

import (
	"encoding/json"
	"net/http"
)

// HTTPEnvelopeHeader is the default HTTP header name for the protocol envelope.
const HTTPEnvelopeHeader = "X-7h3-Envelope"

// VerifyFailReason describes why HTTP envelope verification failed.
type VerifyFailReason string

const (
	ReasonMissingHeader    VerifyFailReason = "missing_header"
	ReasonMalformed        VerifyFailReason = "malformed"
	ReasonUnknownSender    VerifyFailReason = "unknown_sender"
	ReasonInvalidSignature VerifyFailReason = "invalid_signature"
	ReasonExpired          VerifyFailReason = "expired"
)

// KeyRegistry looks up public keys by sender ID.
type KeyRegistry interface {
	GetPublicKey(senderID string) (string, bool)
}

// StaticKeyRegistry is a simple in-memory key registry.
type StaticKeyRegistry struct {
	Keys map[string]string
}

// GetPublicKey returns the public key for the given sender ID.
func (r *StaticKeyRegistry) GetPublicKey(senderID string) (string, bool) {
	k, ok := r.Keys[senderID]
	return k, ok
}

// HTTPVerifyResult is the result of verifying an HTTP envelope.
type HTTPVerifyResult struct {
	OK       bool
	Envelope *ProtocolEnvelope
	Reason   VerifyFailReason
	Detail   string
}

// headerName returns the header name to use (default: HTTPEnvelopeHeader).
func resolveHeaderName(names []string) string {
	if len(names) > 0 && names[0] != "" {
		return names[0]
	}
	return HTTPEnvelopeHeader
}

// VerifyHTTPEnvelope extracts and verifies the protocol envelope from an HTTP request.
func VerifyHTTPEnvelope(r *http.Request, registry KeyRegistry, headerName ...string) HTTPVerifyResult {
	hdr := resolveHeaderName(headerName)
	val := r.Header.Get(hdr)
	if val == "" {
		return HTTPVerifyResult{OK: false, Reason: ReasonMissingHeader, Detail: "header " + hdr + " not present"}
	}

	var env ProtocolEnvelope
	if err := json.Unmarshal([]byte(val), &env); err != nil {
		return HTTPVerifyResult{OK: false, Reason: ReasonMalformed, Detail: err.Error()}
	}

	// Check expiry
	diags := ValidateEnvelope(env)
	for _, d := range diags {
		if d.Message == "Message TTL expired" {
			return HTTPVerifyResult{OK: false, Envelope: &env, Reason: ReasonExpired, Detail: d.Message}
		}
	}

	// Check sender
	if env.Header.Sender == "" {
		return HTTPVerifyResult{OK: false, Envelope: &env, Reason: ReasonMalformed, Detail: "missing sender"}
	}

	pubKey, ok := registry.GetPublicKey(env.Header.Sender)
	if !ok {
		return HTTPVerifyResult{OK: false, Envelope: &env, Reason: ReasonUnknownSender, Detail: "no key for sender: " + env.Header.Sender}
	}

	if env.Signature == nil {
		return HTTPVerifyResult{OK: false, Envelope: &env, Reason: ReasonInvalidSignature, Detail: "no signature"}
	}

	var valid bool
	var verifyErr error
	switch env.Signature.Alg {
	case "ED25519":
		valid, verifyErr = VerifyEnvelopeEd25519(env, pubKey)
	case "HS256":
		valid, verifyErr = VerifyEnvelopeHmac(env, pubKey)
	default:
		return HTTPVerifyResult{OK: false, Envelope: &env, Reason: ReasonInvalidSignature, Detail: "unknown alg: " + env.Signature.Alg}
	}

	if verifyErr != nil {
		return HTTPVerifyResult{OK: false, Envelope: &env, Reason: ReasonInvalidSignature, Detail: verifyErr.Error()}
	}
	if !valid {
		return HTTPVerifyResult{OK: false, Envelope: &env, Reason: ReasonInvalidSignature, Detail: "signature mismatch"}
	}

	return HTTPVerifyResult{OK: true, Envelope: &env}
}

// SignHTTPRequest signs the envelope and returns the header name and JSON value to set.
func SignHTTPRequest(env ProtocolEnvelope, privateKey string, headerName ...string) (name, value string, err error) {
	signed, err := SignEnvelopeEd25519(env, privateKey)
	if err != nil {
		return "", "", err
	}

	b, err := json.Marshal(signed)
	if err != nil {
		return "", "", err
	}

	return resolveHeaderName(headerName), string(b), nil
}

// Middleware returns an http.Handler that verifies the protocol envelope before passing to next.
func Middleware(registry KeyRegistry, next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		result := VerifyHTTPEnvelope(r, registry)
		if !result.OK {
			http.Error(w, string(result.Reason)+": "+result.Detail, http.StatusUnauthorized)
			return
		}
		next.ServeHTTP(w, r)
	})
}

File: sdk/rust/src/http.rs
//! HTTP binding for 7h3 Protocol — signs and verifies per-request envelopes.
//!
//! The signed envelope rides in the `x-7h3-envelope` request header.

use crate::{
    create_envelope, sign_envelope_ed25519, validate_envelope, verify_envelope_ed25519,
    ProtocolBody, ProtocolEnvelope,
};
use std::collections::HashMap;

pub const DEFAULT_HEADER: &str = "x-7h3-envelope";

/// Reason an HTTP envelope verification failed.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum VerifyFailReason {
    MissingHeader,
    MalformedEnvelope(String),
    UnknownSender(String),
    InvalidSignature,
    TtlExpired,
}

impl std::fmt::Display for VerifyFailReason {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::MissingHeader => write!(f, "missing-header"),
            Self::MalformedEnvelope(d) => write!(f, "malformed-envelope: {d}"),
            Self::UnknownSender(s) => write!(f, "unknown-sender: {s}"),
            Self::InvalidSignature => write!(f, "invalid-signature"),
            Self::TtlExpired => write!(f, "ttl-expired"),
        }
    }
}

/// Simple in-memory key registry mapping sender_id → base64url-encoded SPKI public key.
pub struct StaticKeyRegistry {
    keys: HashMap<String, String>,
}

impl StaticKeyRegistry {
    pub fn new(keys: HashMap<String, String>) -> Self {
        Self { keys }
    }

    pub fn get_public_key(&self, sender_id: &str) -> Option<&str> {
        self.keys.get(sender_id).map(|s| s.as_str())
    }
}

/// Input for signing an outbound HTTP request.
pub struct EnvelopeInput {
    pub sender: String,
    pub recipient: Option<String>,
    pub ttl_ms: i64,
    pub body: ProtocolBody,
    pub key_id: String,
}

fn now_ms() -> i64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as i64
}

/// Verify a 7h3 envelope from HTTP headers (generic over header map type).
/// `headers` is a flat map of lowercase header names → values.
pub fn verify_http_envelope(
    headers: &HashMap<String, String>,
    registry: &StaticKeyRegistry,
    header_name: Option<&str>,
    strict_ttl: bool,
) -> Result<ProtocolEnvelope, VerifyFailReason> {
    let hname = header_name.unwrap_or(DEFAULT_HEADER);
    let raw = headers
        .get(hname)
        .or_else(|| headers.get(&hname.to_lowercase()))
        .ok_or(VerifyFailReason::MissingHeader)?;

    let envelope: ProtocolEnvelope = serde_json::from_str(raw)
        .map_err(|e| VerifyFailReason::MalformedEnvelope(e.to_string()))?;

    let sig = envelope
        .signature
        .as_ref()
        .ok_or_else(|| VerifyFailReason::MalformedEnvelope("missing signature".into()))?;

    if strict_ttl {
        let now = now_ms();
        let diags = validate_envelope(&envelope, Some(now));
        if diags.iter().any(|d| d.level == "error") {
            return Err(VerifyFailReason::TtlExpired);
        }
    }

    let sender = &envelope.header.sender;
    match sig.alg.as_str() {
        "ED25519" => {
            let pub_key = registry
                .get_public_key(sender)
                .ok_or_else(|| VerifyFailReason::UnknownSender(sender.clone()))?;
            let valid = verify_envelope_ed25519(&envelope, pub_key)
                .map_err(|e| VerifyFailReason::MalformedEnvelope(e))?;
            if !valid {
                return Err(VerifyFailReason::InvalidSignature);
            }
        }
        "HS256" => {
            let key_id = &sig.key_id;
            return Err(VerifyFailReason::UnknownSender(format!(
                "HMAC key {key_id}: use StaticHmacRegistry"
            )));
        }
        alg => {
            return Err(VerifyFailReason::MalformedEnvelope(format!(
                "unsupported alg: {alg}"
            )));
        }
    }

    Ok(envelope)
}

/// Sign an outbound HTTP request — returns the header name and JSON-serialized envelope value.
pub fn sign_http_request(
    input: &EnvelopeInput,
    private_key: &str,
    header_name: Option<&str>,
) -> Result<(String, String), Box<dyn std::error::Error>> {
    let now = now_ms();
    let envelope = create_envelope(
        &input.sender,
        input.recipient.as_deref(),
        &input.body.intent,
        &input.body.content,
        input.body.capability.as_deref(),
        input.body.correlation_id.as_deref(),
        now,
        input.ttl_ms,
    );
    let signed = sign_envelope_ed25519(&envelope, private_key, &input.key_id)
        .map_err(|e| -> Box<dyn std::error::Error> { e.into() })?;
    let value = serde_json::to_string(&signed)?;
    Ok((header_name.unwrap_or(DEFAULT_HEADER).to_string(), value))
}
